teps/0140-producing-results-in-matrix.md

```diff
@@ -0,0 +1,702 @@
+     |              |              |             |               |                |              |          |          |
+     v              v              v             v               v                v              v          v          v
+report-url-0   report-url-1   report-url-2   report-url-3    report-url-4    report-url-5  report-url-6  report-url-7 report-url-8
+```
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: matrix-testing-platform-and-browsers
+spec:
+  tasks:
+    - name: test-platforms-and-browsers
+      matrix:
+        params:
+          - name: platform
+            value:
+              - linux
+              - mac
+              - windows
+          - name: browser
+            value:
+              - chrome
+              - safari
+              - firefox
+      taskSpec:
+        params:
+          - name: platform
+            type: string
+          - name: browser
+            type: string
+        results:
+          - name: report-url
+            type: string
+         steps:
+          - name: produce-report-url
+            image: alpine
+            script: |
+               echo "Running tests on $(params.platform)-$(params.browser)"
+               echo -n "https://api.example/get-report/$(params.platform)-$(params.browser)" | tee $(results.report-url.path)
+```
+
+### API Change
+
+With this proposal, we add a `ResultsCache` to `ResolvedPipelineTask` that enables caching of `Results` from `Matrixed` `PipelineTasks` in order to prevent resolving the result references for the referenced `Matrixed PipelineTask`  on every single reconcile loop.
+
+```go
+t.ResultsCache[result.Name] = []string{result.Value.StringVal}
+```
+```go
+type ResolvedPipelineTask struct {
+	TaskRunNames []string
+	TaskRuns     []*v1.TaskRun
+	CustomTask     bool
+	CustomRunNames []string
+	CustomRuns     []*v1beta1.CustomRun
+	PipelineTask   *v1.PipelineTask
+	ResolvedTask   *resources.ResolvedTask
+	ResultsCache   map[string][]string
+}
+```
+## Design
+
+### Support String Results produced by a matrix, Aggregated as An Array during Fanning Out - Referencing using indexes
+
+Since the ordering of a fanned out matrix is predeterministic due to sorting the combinations, we can reference results produced from different TaskRuns using indexing using the syntax  `$(tasks.<pipelinetask-name>.results.<result-name>[INDEX])`.  This approach will work for any matrix regardless if it has matrix params and/or include params.  A con to this approach is that users will need to know in advance how the order of the TaskRuns is determined and figure out what the order will be ahead of time if users need to reference or filter for a specific result. At the pipeline definition time, the parameters are unknown. Another con is that it might become exponentially more challenging as you scale with more combinations, making it less user friendly.
```

Comment:
I'm concerned about array indexing into matrixed results for a few reasons:
- UX issues: the user needs to run the pipeline to determine the correct indices to use. If they rename or change params, they will need to rewrite their pipeline.
- Pipelines will not work for different input parameters. This conflicts with our reusability design principle (https://github.com/tektoncd/community/blob/main/design-principles.md#reusability) which states that "At run time (i.e. when invoking a Pipeline or Task via PipelineRun or TaskRun), users should be able to control execution as needed by their context without having to modify Tasks and Pipelines".
- We cannot evolve the implementation of matrix. Deterministic ordering of taskruns was not intended to be a behavior users should rely on (see discussion [here](https://github.com/tektoncd/pipeline/pull/6312#discussion_r1131175378)) and in fact, I think we should allow ordering to be randomized so that users don't come to depend on it. (This is what Go chose to do for the order of iteration over map keys, for a very similar reason: https://go.dev/doc/go1#iteration)

Reply:
For now, I have descoped this TEP to only support emitting and consuming the entire array of results produced by fanned out TaskRuns as this is what everyone currently agrees on and marked this TEP `implementable` so I can continue to make progress, while we continue the discussion in the comments section and in the API working group around handling referencing results for a singular or specific combination(s) of params. 

@lbernick @pritidesai @dibyom @Yongxuanzhang @vdemeester @afrittoli